src/gadget/sinsemilla/chip.rs

```diff
@@ -0,0 +1,240 @@
+    _marker_f: PhantomData<F>,
+    _marker_c: PhantomData<C>,
+}
+
+impl<F: FieldExt, C: CurveAffine<Base = F>> SinsemillaChip<F, C> {
+    /// Configures this chip for use in a circuit.
+    pub fn configure(meta: &mut ConstraintSystem<F>, k: usize, rounds: usize) -> SinsemillaConfig {
+        // Columns required by this chip:
+        // Fixed column for Sinsemilla selector
+        let sinsemilla = meta.fixed_column();
+        let sinsemilla_cur = meta.query_fixed(sinsemilla, Rotation::cur());
+
+        let x_a = meta.advice_column();
+        let z = meta.advice_column();
+        let lambda1 = meta.advice_column();
+        let lambda2 = meta.advice_column();
+        let x_p = meta.advice_column();
+
+        // m_i = z_{i + 1} - (z_i * 2^k)
+        let z_cur = meta.query_advice(z, Rotation::cur());
+        let z_next = meta.query_advice(z, Rotation::next());
+        let m = z_next - z_cur * F::from_u64((1 << k) as u64);
+
+        // y_a = (1/2) ⋅ (lambda1 + lambda2) ⋅ (x_a - (lambda1^2 - x_a - x_p))
+        let lambda1_cur = meta.query_advice(lambda1, Rotation::cur());
+        let lambda2_cur = meta.query_advice(lambda2, Rotation::cur());
+        let x_a_cur = meta.query_advice(x_a, Rotation::cur());
+        let x_p_cur = meta.query_advice(x_p, Rotation::cur());
+        let y_a_cur = (lambda1_cur.clone() + lambda2_cur.clone())
+            * (x_a_cur.clone()
+                - (lambda1_cur.clone() * lambda1_cur.clone() - x_a_cur.clone() - x_p_cur.clone()))
+            * F::from_u64(2).invert().unwrap();
+
+        // y_p = y_a - lambda1 ⋅ (x_a - x_p)
+        let y_p = y_a_cur.clone() - lambda1_cur.clone() * (x_a_cur.clone() - x_p_cur.clone());
+
+        let (x_p_init, y_p_init) = get_s_by_idx::<F, C>(0).to_affine().get_xy().unwrap();
+
+        let lookup_table = GeneratorTable::configure::<F, C>(
+            meta,
+            k,
+            sinsemilla_cur.clone() * m
+                + (Expression::Constant(F::one()) - sinsemilla_cur.clone()) * F::zero(),
+            sinsemilla_cur.clone() * x_p_cur.clone()
+                + (Expression::Constant(F::one()) - sinsemilla_cur.clone()) * x_p_init,
+            sinsemilla_cur.clone() * y_p
+                + (Expression::Constant(F::one()) - sinsemilla_cur.clone()) * y_p_init,
+        );
+
+        let lambda1_next = meta.query_advice(lambda1, Rotation::next());
+        let lambda2_next = meta.query_advice(lambda2, Rotation::next());
+        let x_a_next = meta.query_advice(x_a, Rotation::next());
+        let x_p_next = meta.query_advice(x_p, Rotation::next());
+        let y_a_next = (lambda1_next.clone() + lambda2_next)
+            * (x_a_next.clone()
+                - (lambda1_next.clone() * lambda1_next - x_a_next.clone() - x_p_next))
+            * F::from_u64(2).invert().unwrap();
+
+        // Sinsemilla gate
+        meta.create_gate("Sinsemilla", |_| {
+            // λ_{2,i}^2 − x_{A,i+1} −(λ_{1,i}^2 − x_{A,i} − x_{P,i}) − x_{A,i} = 0
```

Comment:
From pairing with @ebfull:

These need to be two separate gates, rather than combining them into one, because they need to both be equal to zero independently. Otherwise, one term from one of the expressions might be able to cancel out a nonzero result from the other.